src/trace/implementations/merge_batcher.rs

```diff
@@ -179,81 +143,235 @@ impl<D: Ord, T: Ord, R: Semigroup> MergeSorter<D, T, R> {
+
+        // Step 1: Canonicalize each chain by adjacent blocks that combined fit into a single block.
+        for chain in &mut self.queue {
+            let mut target: Vec<(Antichain<T>, Vec<_>)> = Vec::with_capacity(chain.len());
+            for (frontier, block) in chain.drain(..) {
+                if target.last().map_or(false, |(_, last)| {
+                    last.len() + block.len() <= Self::buffer_size()
+                }) {
+                    // merge `target.last()` with `block`
+                    let (last_frontier, last) = target.last_mut().unwrap();
+                    for item in block.into_iter() {
+                        last_frontier.insert_ref(&item.1);
+                        last.push(item);
+                    }
+                } else {
+                    target.push((frontier, block));
+                }
+            }
+            *chain = target;
+        }
+
+        // Step 2: Sort queue by chain length. Depending on how much we extracted,
+        // the chains might be mis-ordered.
+        self.queue.sort_by_key(|chain| std::cmp::Reverse(chain.len()));
+
+        // Step 3: Merge chains that are within a power of two.
+        let mut index = self.queue.len().saturating_sub(1);
+        while index > 0 {
+            if self.queue[index-1].len() / 2 < self.queue[index].len() {
+                // Chains at `index-1` and `index` are within a factor of two, merge them.
+                let list1 = self.queue.remove(index-1);
+                let list2 = std::mem::take(&mut self.queue[index-1]);
+                self.queue[index-1] = self.merge_by(list1, list2);
+            }
+            index -= 1;
         }
-        self.queue_push(list);
+
+        self.account(self.queue.iter().flatten(), 1);
     }
 
-    #[inline(never)]
-    pub fn finish_into(&mut self, target: &mut Vec<Vec<(D, T, R)>>) {
-        while self.queue.len() > 1 {
-            let list1 = self.queue_pop().unwrap();
-            let list2 = self.queue_pop().unwrap();
-            let merged = self.merge_by(list1, list2);
-            self.queue_push(merged);
+    /// Extract all data that is not in advance of `upper`. Record the lower bound of the remaining
+    /// data's time in `frontier`.
+    fn extract_into(
+        &mut self,
+        upper: AntichainRef<T>,
+        frontier: &mut Antichain<T>,
+    ) -> Vec<Vec<(D, T, R)>> {
+        // Flush pending data
+        crate::consolidation::consolidate_updates(&mut self.pending);
+        self.flush_pending();
+
+        let mut keep = self.empty();
+        let mut keep_frontier = Antichain::new();
+        let mut ship = self.empty();
+        let mut ship_list = Vec::default();
+
+        self.account(self.queue.iter().flatten(), -1);
+
+        // Walk all chains, separate ready data from data to keep.
+        for mut chain in std::mem::take(&mut self.queue).drain(..) {
+            let mut block_list = Vec::default();
```

Comment:
Said aloud: "could this be `ship_list`?". Maybe not, but the name `block_list` is not very specific in this context (many lists of blocks here).